Improvements for regrid support for combined models

DougBurke opened the pull request:
I've reworked how this is done, which has lead to bigger changes. I haven't yet updated the documentation to reflect this.

# Summary

Improve behavior of the `regrid` method when called with "complex" models (these with multiple components). Fix #1802 

# Details

Allows UnaryOpModels to be regridded and fixes #1802 (in a way that works for both BinaryOp and UnaryOp models). It is only a small part of the changes needed to get the regrid code to handle the cases we care about.

Adding support for UnaryOpModel:

- change from ArithmeticModel to RegriddableModel as a parent class (all that RegriddableModel does is add the `regrid` method)
- add a `regrid` method that follows the approach of BinaryOpModel, which delegates the choice to the first valid component it finds (this is easier than the binary case as we only have one "part" to check, but it can itself be a complex model so we use the same approach as BinaryOpModel and just look for the `parts` attribute rather than using a specific `is instance` check [*]

Fixing #1802 for BinaryOpModel:

The issue is discussed in #1802, which I repeat here. The problem was that the `regrid` method would look at the two components (`lhs`, `rhs`) via the `parts` attribute. It then tries to find the first class which has a `regrid` method, and applies that class method to `self`. The trouble is when one of these two parts is itself a `BinaryOpModel`, because then it will find it's own code and call it, which ends up in an infinite loop, and hence `RecursionError`. The solution I came up with was to recognize that if a part has a `parts` attribute then we need to recurse through them (which can then recurse if it's a highly-nested expression) until we fit an object with a `regrid` attribute we can call.

Two possible alternatives are

- we can do this recursing when we create the `BinaryOpModel` instance and then store it via a `_regrid` field (for instance), which would make `regard` a lot cleaner and would likely stop so much inspection (since at each level of finding a `*OpModel` we know that will already have a `_regrid` setting). I only came up with this idea after submitting the code, and have decided not to pursue it since a) this code seems to work, b) there's no guarantee there isn't little daemons to deal with in this approach, and c) it means all users pay the price, eve if they never call `regard` (although the costs of all approaches is not expected to be huge)

- an alternative is that we could use the `ndim` attribute of the model to decide whether we call `RegriddableModel1D` or `RegriddableModel2D`. This would be **much** simpler but it combines concerns that should be separate, and would not allow sub-classing of these models (or using an alternative method that provides the same `regrid` interface).

[*] An earlier commit, where we only had support for BinaryOpModel, used is instance, but that was on purpose there as at that time it wasn't valid for UnaryOpModel 

# Background

For those that have not really looked at the regrid code before, I apologize. This change is pretty small, but it may be useful to have a larger-picture view. It also touches on elements I don't want to actually address in this particular PR.

We have the `RegriddableModel` class which adds the `regrid` "interface", by which we can ask for a new model that acts as the original model does but it will evaluate the model on the grid sent to `regrid` and will then "rebin" or interpolate onto whatever user grid is given. The actual implementation is done in`RegriddableModel1D` or `RegriddableModel2D`. It was originally restricted to single models (e.g. gauss1d) and not composite models (e.g. gauss1d + powlaw1d), but this is rather restrictive [**]. In #919 Dan added support for calling `regrid` for a "binary" model - that is `BinaryOpModel`) based on initial prototyping done by me, but it was not correct as discussed in #1802. We want to fix this, and to add support for `regrid` to the `UnaryOpModel` class, as this is the other main way to combine models (much-less-used than `BinaryOpModel`). So, the work here is that for a `regrid` call to a "container" class (`BinaryOpModel` / `UnaryOpModel`), how do we find out what `regrid` method we want from the contained elements (via the `parts` array, which comes from the `CompositeModel` class)?

As an aside, the logic that `regrid` uses can be done by any user, so you don't technically need to have a model that provides the `regrid` method. So, we could just drop the idea of being able to regrid a composite model and tell users that they have to do it manually. I don't think this is a good idea, but it is something we could do. However, we still need to be able to handle regridding composite models (even if not automatically), so there are still going to be follow-up PRs!

[**] I claim that it only really makes sense to regrid a convolution-style model (since the only reason you want data from a different grid is because you are going to pass it to some model to "reprocess" the data into the energy range you care about), but I may be missing a use case, and even if I was right we could have something like `convmodel1(...) + convmodel2(...)` where you want both to use this fancy-grid-extension scheme, or just regrid the whole model as it's too awkward to identify just the convolution component.

DougBurke committed:
Tests: add tests of issue #1802

DougBurke committed:
regrid: fix #1802

The problem was that when regrid was called on a BinaryOpModel
we just took the first component that had a regrid method, but
if this component was itself a BniaryOpModel - such as in

    mdl1 * mdl2 * mdl3

which turns into

    BinaryOpModel(*, BinaryOpModel(*, mdl1, mdl2), mdl3)

and so the regrid check will be on BinaryOpModel(*, mdl1, mdl2),
which has a regrid call but calls itself, leading to no-end of
recursion.

The solution is to explicitly handle BinaryOpModel cases and
hunt through the "full tree" for the first model that has a
regrid call (and is not a BinaryOpModel). The search is left-biased,
but there's no obvious benefit to any other approach.

Some alternatives include:

- explicitly storing the regrid "class" when creating the
  BinaryOpModel (probably in a _regrid field) so we can just
  identify it when calling regrid()

  this should work, but makes everyone pay for this support, when
  we can just do it when we actually need it

- we could use the ndim field for the model, and just pick
  RegriddableModel1D or RegriddableModel2D directly, which would
  be a lot easier, but it would not be extensible.

One concern is that this probably doesn't handle UnaryOpModel cases
well, but they are currently unsupported.

DougBurke committed:
regrid: support UnaryOpModel

Allow UnaryOpModel instances to be regridded.

DougBurke committed:
Tests: check UnaryOpModel regrid handling

hamogu commented:
I have a conceptual question. I guess I should be able to figure it out from the code, but it might be easier to just ask: Assuming we have two model components (say, in a `BinaryOpModel`) and one of the msupports regridding, and the other one does not. What does that mean? Can we regrid the combined model, or not?

Put it another way, I'm thinking of the complex model expression where "basic" models are leaves and `BinaryOpModel` or `UnaryOpModel` or `ConvolutionModel` are nodes. Now, does the entire tree fail to regrid, if there is just one node or leaf that cannot regrid? Or do we identify sub-trees that can regrid, evaluate all parts of that sub-tree on the new grid, and then, inrerpolate/rebin/whatever on the top node of that subtree (so that the output of the subtree is then on the "normal" (not regridded) grid, so that that normal grid can then be passed up into the next higher level (which may not support subgridding)?

Last, there are very few models that cannot support subgridding (there might bem ore where it's not been implemented yet, but that in principle can support that), right? All that comes to mind right now is a `TableModel` with x-axis set to `None`. I'm just trying to understand possible use cases and the fundamental limits.

Why am I asking all this? The code change itself is conceptually simple, but I have not looked (or inderstood) the regrid code before being asked to review this PR and I want to make sure that I'm not approving something that locks us into a solution which I later find out to not agree with what I believe to be the best.

DougBurke commented:
Good question. At the moment we don't really have a way to say "this model can be regridded and this model can not". We could say that only models that derive from the `RegriddableModel` class can be regridded, and then make the *OpModel classes either

- 1. fail when asked to regrid a composite model that does not handle this
- 2. try something really fancy and only regrid those components that support it, and not the others

If we went this way I'd argue for 1, not 2. Thinking about it I think 2 wouldn't be that hard, as it can be handled by over-riding the `calc` method of the `*OpModel` class, but I wonder if you'd get different results if you wrote models differently - e.g. `m1 * (m2 + m3)` vs `m1 * m2 + m1 * m3` when `m1` is not regriddable.

I do not know how many models we have that do not support regridding - I can only think of the table model case when you haven't given an independent axis (which is a highly-specialized case, and has issues if you filter the data, and ideally would be handled separately from the general table-model code, but that's for a different day), and is the problem case you identified too, but I have not done a deep dive on this.

At the moment I am assuming we do not require the `RegriddableModel` superclass, but I'll have a think about this.  One issue we have is that I don't think we've properly gone and updated `ArithmeticModel` to `RegriddableModel1D` (or `2D`) in all our models.

One thing I have in my mind is that this regrid support is only really relevant with a convolution model, otherwise what's the point. That is, we need a model like `fancy(m1 + m2 * m3)` where we want to evaluate `m1 + m2 * m3` on a different grid, use this when convolving it with whatever `fancy` does, and then regridding onto the original grid. In this case we would only need to add `regrid` support to the convolution model and forget all this complication. However, we could end up with `fancy1(some-model-expression) + fancy2(a-different-model-expression)` and want to regrid both `fancy1` and `fancy2` [or have a complicated expression where the convolution is somewhere within it] and just don't want to bother with identifying just the component to change. Also, maybe we do want to be able to mark a model as "un-regriddable", in which case the "only bother with convolution models" approach loses out.

I am not sure this is particularly helpful though!

hamogu commented:
This is a bit helpful, in the sense that it clarifies for me that we might need to think a bit harder on what we want. 

What's the difference between "regrid" and "evaluate on some grid"? Leaving aside the odd "Table model with no x axis" (which is needed in some cases, but clearly it's an odd case), all model *can* can be evaluated on different grids - you just need to set an ignore/notice and the input for x changes. How is that different from regridding? The "lower level" models just receive an array for x and return model values for that; that should work even without a regrid method, right?

So maybe the way to do it is to add the regrid method **only** to models that *cause* regridding to be used such as convolution models, i.e. make regrid a method for "call calc of the lower level models, but with this x grid"? So, regrid would then be something like:
```Python
def regrid(self, x, **kwargs):
    xbig = np.concatenate([
                     np.array([.7, .8, .9]) * x.min(),
                     x,
                     np.array([1.1, 1.2, 1.3]) * x.max()])
    ybig = self.calc(xbig, **kwargs)
    return np.interpolate(x, xbig, ybig)
```

All the lower level models (incl `UnOPModel` and `BinaryOpModel`) would then not need a regrid method because they are just called using the normal `calc` method for some value of x.

In fact, maybe there should not be a `regrid` method at all, but the `calc` method of models that require regridding simply do that as part of their normal `calc` method and then call lower-level models with an expanded grid. The user would not notice the difference.

This runs into problems with models like `fancy1(fancy2(model1 + model2))` in that fancy1 will just call fancy2 with a new grid, so how does fancy2 know of that grid is already regridded? It doesn't so it would regrid again, making the grid even wider and causing the result to be interpolated/rebinned twice with possibly loss of precision. But a-priory it's not clear how to solve this problems anyway - fanc1 and fancy might have different requirements on regridding, e.g. one adds only on the left of the x array and one only one the right. The scheme would work, it might just not be ideal.

Also, I would argue that those situations are rare, and we could pass down a flag "no more regridding" to lower levels.

DougBurke commented:
So, we already have the existing regrid implementation which is being used in prodiuction (for instance the Chandra Source Catalog), so we can't just break things. I believe the use that the CSC makes is only for 2D models [*], and as the 1D and 2D cases are technically separate we can do whatever we want for 1D, but I would want to have a consistent scheme/setup. There are also things I don't understand about the 2D case but that is, I believe, an implementation issue, and discussed in #840 and I think not relevant for the larger picture. 

Although I believe the regrid option only makes sense for convolution style models, I am reluctant to restrict it to just those models as

- I could be wrong (I know @anetasie has used the regrid option in use cases I don't understand)
- I want to support something like the XSPEC [energies](https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSenergies.html) command, which you apply to the whole "dataset/grid", and not just the relevant convolution component and lets you do the equivalent of `plot_source()` and see the model evaluated on the user-defined grid (but `plot_model()` shoes the model evaluated on the "data" grid, ie that provided by the RMF) [**]

I'm not sure I fully understand your issue about "what is special about regrid"? Yes, we should in general be able to evaluate a model on a different grid, that doesn't help you when you say "I need it evaluated on this grid, which is enforced by the data [such as the RMF for PHA data] but I want to internally do the analysis on a different grid and then convert to the 'data grid'". So we need some way to do this operation. But perhaps this is a red herring.

It may be that only being able to regrid convolution models is the cleanest approach, but: we still need to be work out what to do with composite models (the "future reference" point I am about to make is relevant there); as I said, it would be useful to know what @anetasie is actually doing to see if that involves convolution; breaking existing code; and whether we might want to easily view a whole model with the "over-ridden" grid rather than the "what we need to compare to data" grid. But I fear that we may be talking past each other so may have to actually get together in July or later to discuss.

I also want to point out for future reference that "regrid" can mean either

- interpolation onto a different grid
- rebin onto a different grid to conserve "flux"

and understanding what we want to do there depends in part on our discussions on the integrate flag.

---

[*] my understanding is that this is for a convolution-style model (a 2D PSF read in from a file being convolved with a model), so it should fit in the "convolution" narrative 

[**] although I want to be able to support something like XSPEC `energies`, I do not know how we could have `plot_source` use the user-defined grid nicely (it may be easy to do, but you'd not be able to switch between that and the user grid), and there are other issues when it comes to what people expect with `plot_model`with our current code - e.g. 
- #822
- #1300 
- #1349

hamogu commented:
I support "plotting on a high resolution grid". My point is that that's not conceptually a regrid (which to me means "evaluate on one grid and then rebin/interpolate?), but just an "evaluate on the a different grid and plot that" - and the latter is something that might need UI work in the `plot_model` function, but not in how the model is evaluated.
Of course, the trouble is getting the units right (flux per energy? per wavelength? per bin - and if so, which bin? energy flux or counts flux?)

So, the question is, for that vision, what's "regrid" actually supposed to do?

(And I fully agree with not breaking existing behavior **and** also being consistent between 1D and 2D, if at all possible. I'm not trying to convince you to break something, I'm trying to understand what it's supposed to do so that I can confidently say "yes, this PR works towards that goal".)

hamogu commented:
@DougBurke: I agree that a quick talk would probably help. 
@anetasie please let us know what you actually use regrid for.

DougBurke committed:
Merge remote-tracking branch 'upstream/main' into regrid-recursion

DougBurke commented:
Hmm. Looking at this again I'm going to try the "detect which regrid class to use when the composite model is created", which is leading me on some interesting digressions into the existing class hierarchy - e.g. `CompositeModel` vs `ArithmeticModel`.

DougBurke committed:
model: rework model classes to better suppport the regrid method

I wanted to support regrid for unary and binary op models by
identifying the class to use for regridding when the object
was created, rather than when the regrid method was called. This
was primarily driven by trying to ensure test coverage of corner
cases from the old code, and realising that it would likely be
better if done when the class was created.

Of course, this it issues:

- I wanted to store the class as a property of the class, and
  make it optional, but the problem is that the existing
  NoNewAttributesAfterInit behavior forbids over-writing a None
  field with a callable (which the class is).

  So instead I now over-write the get_regrid_class() method; the
  default is that it raises a ModelErr but it can be over-written
  in the RegriddableCompositeModel class.

  The idea is that the RegriddableCompositeModel class can just
  find the first element that has a regrid method (that is not
  in itself going to call itself, leading to the "infinite
  loop" that was issue #1802). It does require some care to ensure
  this, and in the back of my head I am still slightly worried
  about it, but the tests pass so it must be right ...

- I stuck the logic in an "adapter" class, that is
  RegriddableCompositeModel to make it cleaner, but this then
  lead me to windering if this class should also inherit from
  RegriddableModel - which would make sense but I have not done
  so here, because of the next issue (i.e. this was one step too
  far).

- the old class hieararchy didn't make much sense:

  - CompositeModel did not extend ArithmeticModel, and so can not
    be easily combined with other models using Unary/BinaryOpModel
  - however a number of classes that derive from CompositeModel
    did also inherit from ArithmeticModel

  So,

  -> I made CompositeModel inherit from ArithmeticModel
  -> removed the ArithmeticModel inheritance of the subclasses of
     CompositeModel that previously did so
  -> used super() to call parent classes

  With this, we have less diamond inheritance, which is likely
  better but it needs some thinking about.

  A few tests did need to be updated for this change - primarily
  because composite models now gain a cache - but those tests are
  ones that just check the existing behavior (i.e. are regression
  tests) rather than functional ones, so I believe it's okay to
  change.

hamogu commented:
@anetasie What is the status of this? 
Last review note says

> @anetasie please let us know what you actually use regrid for.

Can you tell us your use case so that we can decide how to proceed?

hamogu reviewed (commented):
@anetasie ping: What is the status of this?
